src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java

```diff
@@ -104,7 +105,16 @@ public void setDataProvider(DataProvider<T, ?> dataProvider) {
             dataProviderListenerRegistration.remove();
         }
         dataProviderListenerRegistration = dataProvider
-                .addDataProviderListener(event -> reset());
+                .addDataProviderListener(event -> {
+                    if (event instanceof DataChangeEvent.DataRefreshEvent) {
+                        T t = ((DataChangeEvent.DataRefreshEvent<T>) event).getItem();
+                        this.getCheckboxItems().filter(item -> Objects.equals(item.item, t))
```

Comment:
```
T t = ((DataChangeEvent.DataRefreshEvent<T>) event).getItem();
                        this.getCheckboxItems().filter(item -> Objects.equals(item.item, t))
```
This code should instead of verifying that the items equal, check that the item IDs equal by using `getDataProvider().getId(item)`. By default the `getId(T item)` returns the item itself, but there are cases where you want to override that method and have your own way of determining the identity of the object (like using entity ID).

Please also add a test for this